core/pkg/ingress/types.go

```diff
@@ -0,0 +1,236 @@
+	// location contains information about all the annotations were configured
+	// https://k8s.io/ingress/core/blob/master/pkg/ingress/types.go#L83
+	// The backend returns the contents of the configuration file or an error
+	// with the reason why was not possible to generate the file.
+	//
+	// The returned configuration is then passed to test, and then to reload if there is no errors.
+	OnUpdate(*api.ConfigMap, Configuration) ([]byte, error)
+	// BackendDefaults returns the minimum settings required to configure the
+	// communication to endpoints
+	BackendDefaults() defaults.Backend
+	// Info returns information about the ingress controller
+	Info() *BackendInfo
+}
+
+// BackendInfo returns information about the backend.
+// This fields contains information that helps to track issues or to
+// map the running ingress controller to source code
+type BackendInfo struct {
+	// Name returns the name of the backend
+	Name string
+	// Release returns the running version (semver)
+	Release string
+	// Build returns information about the git commit
+	Build string
+	// Repository return information about the git repository
+	Repository string
+}
+
+// Configuration holds the definition of all the parts required to describe
+type Configuration struct {
+	// HealthzURL location of the URI that is used to know health state of the ingress controller
+	HealthzURL         string
+	Backends           []*Backend
+	Servers            []*Server
+	TCPEndpoints       []*Location
+	UPDEndpoints       []*Location
+	PassthroughBackend []*SSLPassthroughBackend
+}
+
+// Backend describes one or more remote server/s (endpoints) associated with a service
+type Backend struct {
+	// Name represents an unique api.Service name formatted as <namespace>-<name>-<port>
+	Name string
+	// This indicates if the communication protocol between the backend and the endpoint is HTTP or HTTPS
+	// Allowing the use of HTTPS
+	Secure bool
+	// Endpoints contains the list of endpoints currently running
+	Endpoints []Endpoint
+}
+
+// SSLPassthroughBackend describes a SSL upstream server configured
+// as passthrough (no TLS termination in the ingress controller)
+type SSLPassthroughBackend struct {
+	Backend
+
+	Host string
+}
+
+// BackendByNameServers sorts upstreams by name
+type BackendByNameServers []*Backend
+
+func (c BackendByNameServers) Len() int      { return len(c) }
+func (c BackendByNameServers) Swap(i, j int) { c[i], c[j] = c[j], c[i] }
+func (c BackendByNameServers) Less(i, j int) bool {
+	return c[i].Name < c[j].Name
+}
+
+// Endpoint describes a kubernetes endpoint in an backend
+type Endpoint struct {
+	// Address IP address of the endpoint
+	Address string
+	// Port number of the TCP port
+	Port string
+	// MaxFails returns the maximum number of check failures
+	// allowed before this should be considered dow.
+	// Setting 0 indicates that the check is performed by a Kubernetes probe
+	MaxFails int
+
+	FailTimeout int
+}
+
+// Server describes a virtual server
```

Comment:
please clarify what a "virtual" server exactly is, or if there's docs somewhere please point to that